pkg/openfeature/client.go

```diff
@@ -514,22 +547,30 @@ func (c Client) ObjectValueDetails(ctx context.Context, flag string, defaultValu
 	return c.evaluate(ctx, flag, Object, defaultValue, evalCtx, *evalOptions)
 }
 
-func (c Client) evaluate(
+func (c *Client) evaluate(
 	ctx context.Context, flag string, flagType Type, defaultValue interface{}, evalCtx EvaluationContext, options EvaluationOptions,
 ) (InterfaceEvaluationDetails, error) {
 	c.logger().V(debug).Info(
 		"evaluating flag", "flag", flag, "type", flagType.String(), "defaultValue", defaultValue,
 		"evaluationContext", evalCtx, "evaluationOptions", options,
 	)
-	evalCtx = mergeContexts(evalCtx, c.evaluationContext, api.evaluationContext) // API (global) -> client -> invocation
+	evalCtx = mergeContexts(evalCtx, c.evaluationContext, api.evaluationContext()) // API (global) -> client -> invocation
+
+	// ensure that the same provider & hooks are used across this transaction to avoid unexpected behaviour
+	api.RLock()
+	provider := api.prvder
+	globalHooks := api.hks
+	api.RUnlock()
```

Comment:
If I comment out api.RLock() and api.RUnlock here, no tests fail.

Ideally, tests would help us ensure this lock isnt removed, but I'm finding this EXTREMELY hard to test in Java.

Reply:
Done now (whitebox though :( )